### PR TITLE
Support dynamic splits in InProcessCreate

### PR DESCRIPTION
**Pull request by tgroh**
Implement splitAtFraction in InProcessCreate. Create can be split
at any index.

Add @Category(RunnableOnService.class) to existing tests.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @dhalperi

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java [hunk 1]
```diff
@@ -204,6 +209,32 @@ public T getCurrent() throws NoSuchElementException {
...
+      public Double getFractionConsumed() {
+        return rangeTracker.getFractionConsumed();
+      }
+
+      @Override
+      public InMemorySource<T> splitAtFraction(double fraction) {
+        long splitPosition = rangeTracker.getPositionForFractionConsumed(fraction);
+        if (rangeTracker.trySplitAtPosition(splitPosition)) {
+          List<byte[]> oldSourceElems =
+              ImmutableList.copyOf(source.allElementsBytes.subList(0, (int) splitPosition));
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: no need to copy -- that's already done in the constructor.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java [hunk 2]
```diff
@@ -160,21 +159,25 @@ public void validate() {}
...
-      public BytesReader() {
-        this.iter = Iterators.peekingIterator(allElementsBytes.iterator());
+      public BytesReader(InMemorySource<T> source) {
+        rangeTracker = new OffsetRangeTracker(0L, source.allElementsBytes.size());
+        this.source = source;
+        index = -1;
       }
 
       @Override
       public BoundedSource<T> getCurrentSource() {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: `synchronized`

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java [hunk 3]
```diff
@@ -204,6 +209,32 @@ public T getCurrent() throws NoSuchElementException {
...
       @Override
       public void close() throws IOException {}
+
+      @Override
+      public Double getFractionConsumed() {
+        return rangeTracker.getFractionConsumed();
+      }
+
+      @Override
+      public InMemorySource<T> splitAtFraction(double fraction) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: `synchronized`

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java [hunk 4]
```diff
@@ -204,6 +209,32 @@ public T getCurrent() throws NoSuchElementException {
...
+        long splitPosition = rangeTracker.getPositionForFractionConsumed(fraction);
+        if (rangeTracker.trySplitAtPosition(splitPosition)) {
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: invert the `if` to simplify code -- if it fails, return null. then un-indent and actuate the split.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java [hunk 5]
```diff
@@ -204,6 +209,32 @@ public T getCurrent() throws NoSuchElementException {
...
+              ImmutableList.copyOf(source.allElementsBytes.subList(0, (int) splitPosition));
+          List<byte[]> newSourceElems =
+              ImmutableList.copyOf(
+                  source.allElementsBytes.subList(
+                      (int) splitPosition, source.allElementsBytes.size()));
+
+          long totalSize = source.totalSize;
+          long mySourceSizeEstimate = (long) (totalSize * fraction);
+          source = new InMemorySource<>(oldSourceElems, mySourceSizeEstimate, source.coder);
+          return new InMemorySource<>(
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: stick to terminology -- primary and residual, not old and new.

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java [hunk 6]
```diff
@@ -127,7 +132,7 @@ public boolean equals(Object o) {
       return myString.equals(((UnserializableRecord) o).myString);
     }
 
-    static class UnserializableRecordCoder extends StandardCoder<UnserializableRecord> {
+    static class UnserializableRecordCoder extends AtomicCoder<UnserializableRecord> {
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: the change to a deterministic coder is desired?

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java [hunk 7]
```diff
@@ -145,17 +150,27 @@ public UnserializableRecord decode(
...
-      public void verifyDeterministic() throws Coder.NonDeterministicException {
-        stringCoder.verifyDeterministic();
-      }
-    }
+    TestPipeline p = TestPipeline.create();
+    DataflowAssert.that(p.apply(create))
+        .containsInAnyOrder(
+            new UnserializableRecord("foo"),
+            new UnserializableRecord("bar"),
+            new UnserializableRecord("baz"));
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: p.run()?

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java [hunk 8]
```diff
@@ -196,4 +211,17 @@ public void testGetDefaultOutputCoderReturnsConstructorCoder() throws Exception
...
+  public void testSplitAtFraction() throws Exception {
+    List<Integer> elements = new ArrayList<>();
+    Random random = new Random();
+    for (int i = 0; i < 50; i++) {
+      elements.add(random.nextInt());
+    }
+    InProcessCreate.InMemorySource<Integer> source =
+        new InMemorySource<>(elements, VarIntCoder.of());
+
+    SourceTestUtils.assertSplitAtFractionExhaustive(source, PipelineOptionsFactory.create());
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: how long does this take to run? 50 elements can be a lot ;)

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java [hunk 9]
```diff
@@ -204,6 +209,32 @@ public T getCurrent() throws NoSuchElementException {
...
+
+      @Override
+      public InMemorySource<T> splitAtFraction(double fraction) {
+        long splitPosition = rangeTracker.getPositionForFractionConsumed(fraction);
+        if (rangeTracker.trySplitAtPosition(splitPosition)) {
+          List<byte[]> oldSourceElems =
+              ImmutableList.copyOf(source.allElementsBytes.subList(0, (int) splitPosition));
+          List<byte[]> newSourceElems =
+              ImmutableList.copyOf(
+                  source.allElementsBytes.subList(
```
Comment (dhalperi) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: I'd consider seeing if you can simplify this logic. BigtableIO is doing a lot more complex operation here but it manages to have simpler, more readable code :)

https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/io/bigtable/BigtableIO.java#L776

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java (hunk 7 above)
Comment (dhalperi) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: there's another above.

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java [hunk 10]
```diff
@@ -74,6 +78,7 @@ public void testConvertsCreate() {
   }
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: p.run()

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java [hunk 11]
```diff
@@ -74,6 +78,7 @@ public void testConvertsCreate() {
...
 
   @Test
+  @Category(RunnableOnService.class)
   public void testConvertsCreateWithNullElements() {
     Create.Values<String> og =
         Create.<String>of("foo", null, "spam", "ham", null, "eggs")
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: p.run()

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: Done

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: Done

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: done

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java (hunk 4 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: Done.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java (hunk 5 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: Done

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java (hunk 6 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: It's not relevant in this context, but lets us delete some code (the coder is incidentally deterministic, fwiw)

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java (hunk 7 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: Fix't

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java (hunk 8 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: On my machine this is slightly less than 1 second. Reduced to 25, which is slightly more than 0.5s, but should still be sufficiently interesting.

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java (hunk 9 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreate.java]: I've removed some construction which wasn't useful, changed the conditionals, and added a bit of symmetry in how the list is split. I could move the splitting into the source proper, but don't think it's a particularly large improvement.

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java (hunk 10 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: Fix't

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java (hunk 11 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/runners/inprocess/InProcessCreateTest.java]: Fix't

**Comment by tgroh** [comment]
Use an OffsetBasedSource for the whole thing.

**Comment by dhalperi** [comment]
LGTM

**Comment by dhalperi** [comment]
@tgroh presumably want to backport this?

**Pull request closed without merging**